File: backend/src/main/java/com/springboot/controller/exam/ResExamRecordController.java
package com.springboot.controller.exam;

import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.springboot.common.BaseResponse;
import com.springboot.common.ErrorCode;
import com.springboot.common.ResultUtils;
import com.springboot.constant.UserConstant;
import com.springboot.exception.BusinessException;
import com.springboot.exception.ThrowUtils;
import com.springboot.model.dto.exam.ResExamRecordAddRequest;
import com.springboot.model.dto.exam.ResExamRecordQueryRequest;
import com.springboot.model.dto.exam.ResExamRecordUpdateRequest;
import com.springboot.model.entity.exam.ResExamRecord;
import com.springboot.model.vo.exam.ResExamRecordVO;
import com.springboot.model.vo.user.LoginPrincipal;
import com.springboot.service.exam.ResExamRecordService;
import com.springboot.service.user.AuthLoginService;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import java.util.List;
import java.util.Map;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.BeanUtils;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/exam/resExamRecord")
public class ResExamRecordController {

    @Resource
    private ResExamRecordService resExamRecordService;

    @Resource
    private AuthLoginService authLoginService;

    @PostMapping("/student/submit")
    public BaseResponse<Map<String, Object>> submitExam(@RequestBody Map<String, Object> requestBody, HttpServletRequest request) {
        LoginPrincipal principal = assertStudent(request);
        return ResultUtils.success(resExamRecordService.submitExam(
                toLong(requestBody.get("examId")),
                principal.getUserId(),
                castMap(requestBody.get("answers"))));
    }

    @GetMapping("/student/result")
    public BaseResponse<Map<String, Object>> getStudentExamResult(@RequestParam Long examId, HttpServletRequest request) {
        LoginPrincipal principal = assertStudent(request);
        return ResultUtils.success(resExamRecordService.getStudentExamResult(examId, principal.getUserId()));
    }

    @GetMapping("/grading/cards")
    public BaseResponse<List<Map<String, Object>>> listExamRecordCards(@RequestParam Long examId, HttpServletRequest request) {
        assertTeacher(request);
        return ResultUtils.success(resExamRecordService.listExamRecordCards(examId));
    }

    @GetMapping("/grading/records")
    public BaseResponse<List<Map<String, Object>>> listStudentAnswerRecords(@RequestParam Long examId, HttpServletRequest request) {
        assertTeacher(request);
        return ResultUtils.success(resExamRecordService.listStudentAnswerRecords(examId));
    }

    @GetMapping("/grading/detail")
    public BaseResponse<Map<String, Object>> getStudentAnswerRecordDetail(@RequestParam Long recordId, HttpServletRequest request) {
        assertTeacher(request);
        return ResultUtils.success(resExamRecordService.getStudentAnswerRecordDetail(recordId));
    }

    @PostMapping("/grading/grade")
    public BaseResponse<Boolean> gradeAnswer(@RequestBody Map<String, Object> requestBody, HttpServletRequest request) {
        assertTeacher(request);
        return ResultUtils.success(resExamRecordService.gradeAnswer(
                toLong(requestBody.get("recordId")),
                toLong(requestBody.get("questionId")),
                toInteger(requestBody.get("score")),
                requestBody.get("comment") == null ? null : String.valueOf(requestBody.get("comment"))));
    }

    @GetMapping("/grading/stats")
    public BaseResponse<Map<String, Object>> getScoreStatistics(@RequestParam Long examId, HttpServletRequest request) {
        assertTeacher(request);
        return ResultUtils.success(resExamRecordService.getScoreStatistics(examId));
    }

    @PostMapping("/add")
    public BaseResponse<Boolean> addResExamRecord(@RequestBody ResExamRecordAddRequest addRequest) {
        if (addRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        ResExamRecord entity = new ResExamRecord();
        BeanUtils.copyProperties(addRequest, entity);
        resExamRecordService.validResExamRecord(entity, true);
        boolean result = resExamRecordService.save(entity);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(true);
    }

    @PostMapping("/delete")
    public BaseResponse<Boolean> deleteResExamRecord(@RequestParam String id) {
        if (StringUtils.isBlank(id)) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        boolean result = resExamRecordService.removeById(id);
        return ResultUtils.success(result);
    }

    @PostMapping("/update")
    public BaseResponse<Boolean> updateResExamRecord(@RequestBody ResExamRecordUpdateRequest updateRequest) {
        if (updateRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        ResExamRecord entity = new ResExamRecord();
        BeanUtils.copyProperties(updateRequest, entity);
        resExamRecordService.validResExamRecord(entity, false);
        boolean result = resExamRecordService.updateById(entity);
        return ResultUtils.success(result);
    }

    @GetMapping("/get/vo")
    public BaseResponse<ResExamRecordVO> getResExamRecordVOById(@RequestParam String id) {
        if (StringUtils.isBlank(id)) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        ResExamRecord entity = resExamRecordService.getById(id);
        ThrowUtils.throwIf(entity == null, ErrorCode.NOT_FOUND_ERROR);
        return ResultUtils.success(resExamRecordService.getResExamRecordVO(entity, null));
    }

    @PostMapping("/list/page")
    public BaseResponse<Page<ResExamRecord>> listResExamRecordByPage(@RequestBody ResExamRecordQueryRequest queryRequest) {
        if (queryRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        long current = queryRequest.getCurrent();
        long size = queryRequest.getPageSize();
        Page<ResExamRecord> page = resExamRecordService.page(new Page<>(current, size), resExamRecordService.getQueryWrapper(queryRequest));
        return ResultUtils.success(page);
    }

    @PostMapping("/list/page/vo")
    public BaseResponse<Page<ResExamRecordVO>> listResExamRecordVOByPage(@RequestBody ResExamRecordQueryRequest queryRequest) {
        if (queryRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        long current = queryRequest.getCurrent();
        long size = queryRequest.getPageSize();
        ThrowUtils.throwIf(size > 50, ErrorCode.PARAMS_ERROR);
        Page<ResExamRecord> page = resExamRecordService.page(new Page<>(current, size), resExamRecordService.getQueryWrapper(queryRequest));
        return ResultUtils.success(resExamRecordService.getResExamRecordVOPage(page, null));
    }

    private LoginPrincipal assertTeacher(HttpServletRequest request) {
        LoginPrincipal principal = authLoginService.getLoginPrincipal(request);
        if (!UserConstant.ROLE_TYPE_TEACHER.equals(principal.getRoleType())) {
            throw new BusinessException(ErrorCode.NO_AUTH_ERROR, "仅教师可访问考试批改接口");
        }
        return principal;
    }

    private LoginPrincipal assertStudent(HttpServletRequest request) {
        LoginPrincipal principal = authLoginService.getLoginPrincipal(request);
        if (!UserConstant.ROLE_TYPE_STUDENT.equals(principal.getRoleType())) {
            throw new BusinessException(ErrorCode.NO_AUTH_ERROR, "仅学生可提交考试");
        }
        return principal;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> castMap(Object value) {
        if (value instanceof Map<?, ?> map) {
            return (Map<String, Object>) map;
        }
        return Map.of();
    }

    private Long toLong(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Number number) {
            return number.longValue();
        }
        return Long.parseLong(String.valueOf(value));
    }

    private Integer toInteger(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Number number) {
            return number.intValue();
        }
        return Integer.parseInt(String.valueOf(value));
    }
}

File: frontend/src/api/resExamRecordController.ts
// @ts-ignore
/* eslint-disable */
import request from "@/request";

/** 此处后端没有提供注释 POST /exam/resExamRecord/add */
export async function addResExamRecord(
  body: API.ResExamRecordAddRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>("/exam/resExamRecord/add", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** 此处后端没有提供注释 POST /exam/resExamRecord/delete */
export async function deleteResExamRecord(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.deleteResExamRecordParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>("/exam/resExamRecord/delete", {
    method: "POST",
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/** 此处后端没有提供注释 GET /exam/resExamRecord/get/vo */
export async function getResExamRecordVoById(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getResExamRecordVOByIdParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseResExamRecordVO>(
    "/exam/resExamRecord/get/vo",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 GET /exam/resExamRecord/grading/cards */
export async function listExamRecordCards(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listExamRecordCardsParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseListMapStringObject>(
    "/exam/resExamRecord/grading/cards",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 GET /exam/resExamRecord/grading/detail */
export async function getStudentAnswerRecordDetail(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getStudentAnswerRecordDetailParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseMapStringObject>(
    "/exam/resExamRecord/grading/detail",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /exam/resExamRecord/grading/grade */
export async function gradeAnswer(
  body: Record<string, any>,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>("/exam/resExamRecord/grading/grade", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** 此处后端没有提供注释 GET /exam/resExamRecord/grading/records */
export async function listStudentAnswerRecords(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listStudentAnswerRecordsParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseListMapStringObject>(
    "/exam/resExamRecord/grading/records",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 GET /exam/resExamRecord/grading/stats */
export async function getScoreStatistics(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getScoreStatisticsParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseMapStringObject>(
    "/exam/resExamRecord/grading/stats",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /exam/resExamRecord/list/page */
export async function listResExamRecordByPage(
  body: API.ResExamRecordQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageResExamRecord>(
    "/exam/resExamRecord/list/page",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /exam/resExamRecord/list/page/vo */
export async function listResExamRecordVoByPage(
  body: API.ResExamRecordQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageResExamRecordVO>(
    "/exam/resExamRecord/list/page/vo",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /exam/resExamRecord/student/submit */
export async function submitExam(
  body: Record<string, any>,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseMapStringObject>(
    "/exam/resExamRecord/student/submit",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /exam/resExamRecord/update */
export async function updateResExamRecord(
  body: API.ResExamRecordUpdateRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>("/exam/resExamRecord/update", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}
